### ISSUE TITLE
FineTuning and ReduceLROnPleateau scheduler fail - optimizer.param_groups

### ISSUE THREAD

**Issue by FlorianMF** [PyTorchLightning/pytorch-lightning#8720]
### Context
I would like to use Finetuning and ReduceLROnPlateau scheduler.

For the finetuning part I use the same [MilestonesFinetuning](https://github.com/PyTorchLightning/pytorch-lightning/blob/052aefc3420eb507cd4eb7256ca452118448cd66/pl_examples/domain_templates/computer_vision_fine_tuning.py#L69) callback as in this example .
I used to instantiate the optimizer in "configure_optimizers()" the same way as in the above example: 
`optimizer(filter(lambda p: p.requires_grad, self.parameters()), **kwargs)`

ReduceLROnPlateau iterates over optimizer.param_groups and min_lrs. The latter is defined at scheduler instantiation. Hence its of length 1.
When unfreezing some layers with the callback, a new param_group is added to the optimizer.
-> len(param_groups) != len(min_lrs). hence it fails.

When simply instantiating the optimizer w/o filtering the parameters the layers are simply made trainable as they are already known to the optimizer.

### Question
Is there a real advantage in filtering on trainable params at optimizer init time, or can I just init as follows: `optimizer(self.parameters(), **kwargs)` ?
what I find interesting about adding param groups is that they are added w/ a lower learning_rate.

### Dummy script
I reconstructed a simple script using some random data, the MilestonesFinetuning callback and the BoringModel adapted to the callback.
With the arg `filter_on_trainable: bool` you can switch between both optimizer initialization options.

```
import torch
from torch import optim
from torch import nn
from torch.optim.lr_scheduler import ReduceLROnPlateau
from torch.optim.optimizer import Optimizer
from torch.utils.data import DataLoader, Dataset

import pytorch_lightning as pl
from pytorch_lightning.callbacks.finetuning import BaseFinetuning


class MilestonesFinetuning(BaseFinetuning):

    def __init__(self, milestones: tuple = (5, 10), train_bn: bool = False):
        super().__init__()
        self.milestones = milestones
        self.train_bn = train_bn

    def freeze_before_training(self, pl_module: pl.LightningModule):
        self.freeze(
            modules=pl_module.feature_extractor, train_bn=self.train_bn
        )

    def finetune_function(
        self, pl_module: pl.LightningModule, epoch: int, optimizer: Optimizer,
        opt_idx: int
    ):
        if epoch == self.milestones[0]:
            # unfreeze 5 last layers
            self.unfreeze_and_add_param_group(
                modules=pl_module.feature_extractor[-5:],
                optimizer=optimizer,
                train_bn=self.train_bn
            )

        elif epoch == self.milestones[1]:
            # unfreeze remaing layers
            self.unfreeze_and_add_param_group(
                modules=pl_module.feature_extractor[:-5],
                optimizer=optimizer,
                train_bn=self.train_bn
            )


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(pl.LightningModule):

    def __init__(self, filter_on_trainable: bool = True):
        """
        Testing PL Module
        Use as follows:
        - subclass
        - modify the behavior for what you want
        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing
        or:
        model = BaseTestModel()
        model.training_epoch_end = None
        """
        super().__init__()
        self.filter_on_trainable = filter_on_trainable
        self.feature_extractor = nn.Sequential(
            torch.nn.Linear(64, 64),
            torch.nn.Linear(64, 32),
            torch.nn.Linear(32, 32),
            torch.nn.Linear(32, 16),
            torch.nn.Linear(16, 16),
            torch.nn.Linear(16, 8),
            torch.nn.Linear(8, 8),
            torch.nn.Linear(8, 4),
        )
        self.fc = torch.nn.Linear(4, 2)

    def forward(self, x):
        return self.fc(self.feature_extractor(x))

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(
            prediction, torch.ones_like(prediction)
        )

    def step(self, x):
        x = self(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        loss_mean = torch.stack([x["x"] for x in outputs]).mean()
        self.log('val_loss', loss_mean)

    def test_step(self, batch, batch_idx):
        output = self(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        parameters = list(self.parameters())
        if self.filter_on_trainable:
            trainable_parameters = list(
                filter(lambda p: p.requires_grad, parameters)
            )
            optimizer = optim.Adam(trainable_parameters, lr=0.1)
        else:
            optimizer = optim.Adam(parameters, lr=0.1)
        # set threshold very high such that reduce_lr is called immediately
        scheduler = ReduceLROnPlateau(optimizer, patience=0, threshold=1e+5)
        return {
            'optimizer': optimizer,
            'lr_scheduler': scheduler,
            'monitor': 'val_loss'
        }

    def train_dataloader(self):
        return DataLoader(RandomDataset(64, 1))

    def val_dataloader(self):
        return DataLoader(RandomDataset(64, 1))


def main():
    callbacks = [MilestonesFinetuning(milestones=(1, 3))]
    model = BoringModel(filter_on_trainable=True)
    trainer = pl.Trainer(callbacks=callbacks)
    trainer.fit(model)


if __name__ == "__main__":
    main()
```
### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by FlorianMF** [comment]
Extracts from the scheduler code:
```
def _reduce_lr(self, epoch):
        for i, param_group in enumerate(self.optimizer.param_groups):
            old_lr = float(param_group['lr'])
            new_lr = max(old_lr * self.factor, self.min_lrs[i])
```

and from `BaseFinetuning`:
```
def unfreeze_and_add_param_group(
        BaseFinetuning.make_trainable(modules)
        params_lr = optimizer.param_groups[0]["lr"] if lr is None else float(lr)
        denom_lr = initial_denom_lr if lr is None else 1.0
        params = BaseFinetuning.filter_params(modules, train_bn=train_bn, requires_grad=True)
        params = BaseFinetuning.filter_on_optimizer(optimizer, params)
        if params:
            optimizer.add_param_group({"params": params, "lr": params_lr / denom_lr})
```

**Comment by icedpanda** [comment]
Hi @FlorianMF, I am facing the same issue using `ReduceLROnPleateau` while doing multi-phases training. Do you find any solution for this by any chance? Cheers

**Comment by FlorianMF** [comment]
Hey @icedpanda, unfortunately no. 
In the end I removed the `filter(lambda ...)` part and just stuck to passing all parameters in.

**Comment by icedpanda** [comment]
@FlorianMF, thanks for the quick replay.  I wonder does optimizer update weights correctly if you pass all parameters to the optimizer while freezing and unfreezing part of the networks during the training?

**Comment by FlorianMF** [comment]
Yes, the freezing-unfreezing works.
The Finetuning callback simply won't add these parameters in a new parameter group. Hence they will have the same learning rate as the already non-frozen instead of the default 1/10th value as implemented in the callback for a new param_group

**Comment by patrontheo** [comment]
The issue is form pytorch I think: [issue](https://github.com/pytorch/pytorch/issues/20997)
